Build issues with GCC v9

While building NEURON with CMake using GCC 9 (or 9.2), I am seeing below errors:


```
$ cmake .. -DNRN_ENABLE_CORENEURON=OFF -DNRN_ENABLE_INTERVIEWS=OFF
....
-- The C compiler identification is GNU 9.2.0
-- The CXX compiler identification is GNU 9.2.0
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/bin/gcc
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
...
--
-- Configured NEURON 7.8
--
-- Some things you can do now:
-- --------------+--------------------------------------------------------------
-- Command       |   Description
-- --------------+--------------------------------------------------------------
-- make install  | Will install NEURON to: /usr/local
--               | Change the install location of NEURON using:
--               |     cmake <src_path> -DCMAKE_INSTALL_PREFIX=<install_path>
-- make docs     | Build the API documentation
-- make uninstall| Removes files installed by make install (todo)
-- --------------+--------------------------------------------------------------
--  Build option | Status
-- --------------+--------------------------------------------------------------
-- Shared        | ON
-- Legacy FR     | ON
-- MPI           | ON
--   INC         | /opt/hpe/hpc/mpt/mpt-2.16/include
--   LIB         | /lib64/libpthread.so
-- Python        | ON
--   EXE         | /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/external-libraries/2018-12-19/linux-rhel7-x86_64/gcc-6.4.0/python-3.6.5-ukuow6gd2f/bin/python
--   INC         | /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/external-libraries/2018-12-19/linux-rhel7-x86_64/gcc-6.4.0/python-3.6.5-ukuow6gd2f/include/python3.6m
--   LIB         | /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/external-libraries/2018-12-19/linux-rhel7-x86_64/gcc-6.4.0/python-3.6.5-ukuow6gd2f/lib/libpython3.6m.so
--   DYNAMIC     | OFF
-- RX3D          | ON
--   OptLevel    | 0
-- Interviews    | OFF
-- CoreNEURON    | OFF
-- --------------+--------------------------------------------------------------
--  See documentation : https://www.neuron.yale.edu/neuron/
-- --------------+--------------------------------------------------------------
--
-- Configuring done
-- Generating done
-- Build files have been written to: /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build
...

$ make VERBOSE=1
/gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/tools/2018-12-19/linux-rhel7-x86_64/gcc-6.4.0/cmake-3.13.0-ivumccos42/bin/cmake -S/gpfs/bbp.cscs.ch/home/kumbhar/tmp/
....
[  6%] Building CXX object src/nrniv/CMakeFiles/nrniv_lib.dir/bbsdirect.cpp.o
cd /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrniv && /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/bin/g++  -DHAVE_CONFIG_H -Dnrniv_lib_EXPORTS -DCABLE=1 -DOOP=1 -DUSEBBS=1 -DUSECVODE=1 -DUSEMATRIX=1 -I/opt/hpe/hpc/mpt/mpt-2.16/include -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/ivoc -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrnoc -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/oc -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrncvode -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrnjava -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrnoc -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrnpython -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/oc -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/parallel -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/sundials -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/sundials/shared -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/gnu -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/memacs -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/mesch -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrncvode -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrnmpi -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrnpython -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/parallel -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sparse -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sparse13 -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sundials -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sundials/cvodes -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sundials/ida -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/sundials/shared -I/gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/external-libraries/2018-12-19/linux-rhel7-x86_64/gcc-6.4.0/python-3.6.5-ukuow6gd2f/include/python3.6m -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/nrniv_lib -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/ivos -I/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/ivos  -fPIC   -std=c++98 -o CMakeFiles/nrniv_lib.dir/bbsdirect.cpp.o -c /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/bbsdirect.cpp
In file included from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/bbsdirect.cpp:3:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:28: error: expected template-name before ‘<’ token
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |                            ^
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:28: error: expected ‘{’ before ‘<’ token
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:28: error: expected unqualified-id before ‘<’ token
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp: In constructor ‘BBSDirect::BBSDirect()’:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:48:27: error: invalid use of incomplete type ‘class KeepArgs’
   48 |  keepargs_ = new KeepArgs();
      |                           ^
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:7: note: forward declaration of ‘class KeepArgs’
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |       ^~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp: In destructor ‘virtual BBSDirect::~BBSDirect()’:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:56:9: warning: possible problem detected in invocation of delete operator: [-Wdelete-incomplete]
   56 |  delete keepargs_;
      |         ^~~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:56:9: warning: invalid use of incomplete type ‘class KeepArgs’
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:7: note: forward declaration of ‘class KeepArgs’
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |       ^~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:56:9: note: neither the destructor nor the class-specific operator delete will be called, even if they are declared when the class is defined
   56 |  delete keepargs_;
      |         ^~~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp: In member function ‘virtual void BBSDirect::save_args(int)’:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:269:11: error: invalid use of incomplete type ‘class KeepArgs’
  269 |  keepargs_->insert(
      |           ^~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:7: note: forward declaration of ‘class KeepArgs’
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |       ^~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:270:3: error: ‘pair’ was not declared in this scope; did you mean ‘std::pair’?
  270 |   pair<const int, bbsmpibuf*>(userid, sendbuf_)
      |   ^~~~
      |   std::pair
In file included from /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/bits/stl_algobase.h:64,
                 from /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/bits/char_traits.h:39,
                 from /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/ios:40,
                 from /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/ostream:38,
                 from /gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/iostream:39,
                 from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/ivos/ivstream.h:4,
                 from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build/src/nrncvode/../../nrnconf.h:370,
                 from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:1,
                 from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/bbsdirect.cpp:3:
/gpfs/bbp.cscs.ch/apps/hpc/jenkins/i_would_like_to_be_deleted---merge-2019-10/deploy/compilers/2019-10-02/linux-rhel7-x86_64/gcc-4.8.5/gcc-9.2.0-tuxi4uvc6t/include/c++/9.2.0/bits/stl_pair.h:208:12: note: ‘std::pair’ declared here
  208 |     struct pair
      |            ^~~~
In file included from /gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/bbsdirect.cpp:3:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:270:8: error: expected primary-expression before ‘const’
  270 |   pair<const int, bbsmpibuf*>(userid, sendbuf_)
      |        ^~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:270:28: error: expected primary-expression before ‘*’ token
  270 |   pair<const int, bbsmpibuf*>(userid, sendbuf_)
      |                            ^
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:270:29: error: expected primary-expression before ‘>’ token
  270 |   pair<const int, bbsmpibuf*>(userid, sendbuf_)
      |                             ^
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp: In member function ‘virtual void BBSDirect::return_args(int)’:
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:279:12: error: incomplete type ‘KeepArgs’ used in nested name specifier
  279 |  KeepArgs::iterator i = keepargs_->find(userid);
      |            ^~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:282:6: error: ‘i’ was not declared in this scope
  282 |  if (i != keepargs_->end()) {
      |      ^
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:282:20: error: invalid use of incomplete type ‘class KeepArgs’
  282 |  if (i != keepargs_->end()) {
      |                    ^~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:7: note: forward declaration of ‘class KeepArgs’
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |       ^~~~~~~~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:284:12: error: invalid use of incomplete type ‘class KeepArgs’
  284 |   keepargs_->erase(i);
      |            ^~
/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/src/nrniv/../parallel/bbsdirectmpi.cpp:35:7: note: forward declaration of ‘class KeepArgs’
   35 | class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
      |       ^~~~~~~~
make[2]: *** [src/nrniv/CMakeFiles/nrniv_lib.dir/bbsdirect.cpp.o] Error 1
make[2]: Leaving directory `/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build'
make[1]: *** [src/nrniv/CMakeFiles/nrniv_lib.dir/all] Error 2
make[1]: Leaving directory `/gpfs/bbp.cscs.ch/home/kumbhar/tmp/nrn/build'
make: *** [all] Error 2
```

This seems to be related to  #358 but I am surprised that this is only appearing with GCC v9.

I am shorty creating PR with the fix.

cc: @anilbey

--- a/src/parallel/bbsclimpi.cpp
+++ b/src/parallel/bbsclimpi.cpp
@@ -37,7 +37,7 @@ struct ltint {
 	}
 };
 
-class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
+class KeepArgs : public std::map<int, bbsmpibuf*, ltint>{};
 
 #endif
 
@@ -274,7 +274,7 @@ void BBSClient::save_args(int userid) {
 #if defined(HAVE_STL)
 	nrnmpi_ref(sendbuf_);
 	keepargs_->insert(
-		pair<const int, bbsmpibuf* >(userid, sendbuf_)
+		std::pair<const int, bbsmpibuf* >(userid, sendbuf_)
 	);
 	
 #endif

--- a/src/parallel/bbsdirectmpi.cpp
+++ b/src/parallel/bbsdirectmpi.cpp
@@ -32,7 +32,7 @@ struct ltint {
 	}
 };
 
-class KeepArgs : public map<int, bbsmpibuf*, ltint>{};
+class KeepArgs : public std::map<int, bbsmpibuf*, ltint>{};
 
 #endif
 
@@ -267,7 +267,7 @@ void BBSDirect::save_args(int userid) {
 #if defined(HAVE_STL)
 	nrnmpi_ref(sendbuf_);
 	keepargs_->insert(
-		pair<const int, bbsmpibuf*>(userid, sendbuf_)
+		std::pair<const int, bbsmpibuf*>(userid, sendbuf_)
 	);
 	
 #endif

--- a/src/parallel/bbssrv2mpi.cpp
+++ b/src/parallel/bbssrv2mpi.cpp
@@ -104,12 +104,12 @@ printf("todo_less_than %d < %d return %d\n", this->id_, w->id_, w1->id_ < w2->id
 	return w1->id_ < w2->id_;
 }
 
-class MessageList : public multimap <const char*, bbsmpibuf*, ltstr>{};
-class PendingList : public multimap <const char*, const int, ltstr>{};
-class WorkList : public map <int, const WorkItem*, ltint>{};
-class LookingToDoList : public set <int, ltint>{};
-class ReadyList : public set<const WorkItem*, ltWorkItem>{};
-class ResultList: public multimap<int, const WorkItem*, ltint>{};
+class MessageList : public std::multimap <const char*, bbsmpibuf*, ltstr>{};
+class PendingList : public std::multimap <const char*, const int, ltstr>{};
+class WorkList : public std::map <int, const WorkItem*, ltint>{};
+class LookingToDoList : public std::set <int, ltint>{};
+class ReadyList : public std::set<const WorkItem*, ltWorkItem>{};
+class ResultList: public std::multimap<int, const WorkItem*, ltint>{};
 #else
 class MessageList {};
 class PendingList {};
@@ -201,7 +201,7 @@ void BBSDirectServer::put_pending(const char* key, int cid) {
 printf("put_pending |%s| %d\n", key, cid);
 #endif
 	char* s = newstr(key);
-	pending_->insert(pair<const char* const, const int>(s, cid));
+	pending_->insert(std::pair<const char* const, const int>(s, cid));
 #endif
 }
 
@@ -233,7 +233,7 @@ void BBSDirectServer::post(const char* key, bbsmpibuf* send) {
 		nrnmpi_bbssend(cid, TAKE, send);
 	}else{
 		MessageList::iterator m = messages_->insert(
-		  pair<const char* const, bbsmpibuf*>(newstr(key), send)
+		  std::pair<const char* const, bbsmpibuf*>(newstr(key), send)
 		);
 		nrnmpi_ref(send);
 	}
@@ -257,7 +257,7 @@ printf("BBSDirectServer::post_todo pid=%d cid=%d send=%p\n", pid, cid, send);
 	if (p != work_->end()) {
 		w->parent_ = (WorkItem*)((*p).second);
 	}
-	work_->insert(pair<const int, const WorkItem*>(w->id_, w));
+	work_->insert(std::pair<const int, const WorkItem*>(w->id_, w));
 #if debug
 printf("work insert %d\n", w->id_);
 #endif
@@ -356,7 +356,7 @@ printf("DirectServer::post_result id=%d send=%p\n", id, send);
 	nrnmpi_ref(send);
 	nrnmpi_unref(w->buf_);
 	w->buf_ = send;
-	results_->insert(pair<const int, const WorkItem*>(w->parent_ ? w->parent_->id_ : 0, w));
+	results_->insert(std::pair<const int, const WorkItem*>(w->parent_ ? w->parent_->id_ : 0, w));
 #endif
 }
 

